### DISCUSSION TITLE
Uncaught Error: A discriminator value for key could not be extracted from all schema options

### DISCUSSION THREAD

**Discussion by TitoSniberb** [colinhacks/zod#2747] [Q&A] [answered]
Having this code:
```ts
const portfolioManagedInternalizationData = z
  .object({ serviceTypeId: z.number().int() })
  .passthrough();
const portfolioAdviceInternalizationData = z
  .object({
    serviceTypeId: z.number().int(),
    adviceTypeId: z.number().int(),
    bankingEntityId: z.union([z.number(), z.null()]).nullish(),
  })
  .passthrough();
const portfolioInternalizationData = z.discriminatedUnion("serviceTypeId", [
  portfolioManagedInternalizationData,
  portfolioAdviceInternalizationData,
]);
```

Why is it showing this error?
`Uncaught Error: A discriminator value for key 'serviceTypeId' could not be extracted from all schema options`

**Comment by JacobWeisenburger** [comment, marked as answer]
1. `z.discriminatedUnion` requires the use of a `z.literal` or `z.enum`. It doesn't work with `z.number`.
1. `z.union([z.number(), z.null()]).nullish()` is the same as `z.number().nullish()`

Try doing something like this:

```ts
const portfolioManagedInternalizationData = z.object( {
    serviceTypeId: z.literal( 'foo' )
} ).passthrough()

const portfolioAdviceInternalizationData = z.object( {
    serviceTypeId: z.enum( [ 'bar', 'baz' ] ),
    adviceTypeId: z.number().int(),
    bankingEntityId: z.number().nullish(),
} ).passthrough()

const portfolioInternalizationData = z.discriminatedUnion( 'serviceTypeId', [
    portfolioManagedInternalizationData,
    portfolioAdviceInternalizationData,
] )

console.log(
    portfolioInternalizationData.safeParse( {
        serviceTypeId: 'foo',
    } ).success
) // true

console.log(
    portfolioInternalizationData.safeParse( {
        serviceTypeId: 'bar',
        adviceTypeId: 1,
        bankingEntityId: null,
    } ).success
) // true

console.log(
    portfolioInternalizationData.safeParse( {
        serviceTypeId: 'baz',
        adviceTypeId: 2,
        bankingEntityId: 42,
    } ).success
) // true
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by arthurvanl** [reply]
Hi Jacob,

I am having a similar issue:

![image](https://github.com/colinhacks/zod/assets/20114845/b174ca50-f3ed-46d4-b51b-f6ec46decf3b)

I do like to use other options that are not provided in the union. Those object must only have the required fields. I tried adding this
![image](https://github.com/colinhacks/zod/assets/20114845/d592f329-467b-4784-85d9-65b1c43fe6c6)
in the discriminated union except that throws the initial error.
